scripts/fast_lint.py

```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env python3
+
+"""Runs the common linters on any files that have changed relative to the main branch."""
+from __future__ import annotations
+
+import argparse
+import concurrent.futures
+import logging
+import os
+import subprocess
+import sys
+import time
+
+from git import Repo
+
+
+def changed_files() -> list[str]:
+    repo = Repo(os.getcwd())
+
+    current_branch = repo.active_branch
+    common_ancestor = repo.merge_base(current_branch, "main")[0]
+
+    return [item.b_path for item in repo.index.diff(common_ancestor)]
+
+
+def run_cmd(cmd: list[str], files: list[str] | bool) -> bool:
+    start = time.time()
+    if not files:
+        logging.info(f"SKIP: `{' '.join(cmd)}`")
+        return True
+
+    if isinstance(files, bool):
+        files = []
+    proc = subprocess.run(cmd + files, text=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+    if proc.returncode == 0:
+        logging.info(f"PASS: `{' '.join(cmd)} {'<FILES>' if files else ''}` in {time.time() - start:.2f}s")
+        if proc.stdout:
+            logging.debug(f"stdout: {proc.stdout}")
+    else:
+        logging.info(f"FAIL: `{' '.join(cmd)} {'<FILES>' if files else ''}` in {time.time() - start:.2f}s")
+        if proc.stdout:
+            logging.info(f"stdout: {proc.stdout}")
+
+    return proc.returncode == 0
+
+
+def filter_ext(files: list[str], ext: list[str]) -> list[str]:
+    return [f for f in files if any(f.endswith(e) for e in ext)]
+
+
+def filter_cpp(files: list[str]) -> list[str]:
+    return [f for f in files if f.endswith(".h") or f.endswith(".hpp") or f.endswith(".c") or f.endswith(".cpp")]
+
+
+def main() -> None:
+    start = time.time()
+    parser = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
+    parser.add_argument(
+        "--log-level",
+        dest="log_level",
+        default="INFO",
+        choices=["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"],
+        help="Set the logging level (default: INFO)",
+    )
+    parser.add_argument("--num-threads", type=int, default=8, help="Number of threads to use (default: 4)")
+    parser.add_argument(
+        "files",
+        metavar="file",
+        type=str,
+        nargs="*",
+        help="File paths. Empty = all files, recursively.",
+    )
+
+    args = parser.parse_args()
+    logging.basicConfig(level=args.log_level)
+
+    script_dirpath = os.path.dirname(os.path.realpath(__file__))
+    root_dirpath = os.path.abspath(f"{script_dirpath}/..")
+    os.chdir(root_dirpath)
+
+    if args.files:
+        files = args.files
+    else:
+        files = changed_files()
+
+    logging.debug("Checking:")
+    for f in files:
+        logging.debug(f"  {f}")
+
+    jobs = [
+        (["pixi", "run", "codegen", "--check"], True),
```

Comment:
This one takes 5-7s for me, which is dangerously close to some threshold as far as pre-commit hook are concerned.

```
❯ python scripts/fast_lint.py 
INFO:root:SKIP: `rustfmt --check`
INFO:root:SKIP: `taplo fmt --check`
INFO:root:SKIP: `pixi run clang-format --dry-run`
INFO:root:PASS: `ruff check --config rerun_py/pyproject.toml <FILES>` in 0.01s
INFO:root:PASS: `typos <FILES>` in 0.04s
INFO:root:PASS: `python scripts/lint.py <FILES>` in 0.10s
INFO:root:PASS: `blackdoc --check <FILES>` in 0.51s
INFO:root:PASS: `black --check --config rerun_py/pyproject.toml <FILES>` in 0.54s
INFO:root:FAIL: `mypy --no-warn-unused-ignore <FILES>` in 0.99s
INFO:root:stdout: scripts/screenshot_compare/build_screenshot_compare.py:84: error: Incompatible types in assignment (expression has type "str", variable has type "Path")  [assignment]
Found 1 error in 1 file (checked 2 source files)

INFO:root:PASS: `pixi run codegen --check ` in 5.70s
INFO:root:Lints failed in 5.73s
```
That's after a few runs to warm things up.

Reply:
(running on my `antoine/screenshot_compare` branch)

Reply:
>as far as pre-commi

Ahh yeah, agreed it's slow for a pre-commit. I have been primarily thinking of this as a pre-push, which I think strikes the right balance. I've now added option to skip arbitrary jobs if you wanted to use for a pre-commit as well.

Reply:
pre-push makes sense, yeah 👍🏻

Reply:
Still sounds too slow for a pre-push for my taste. It's also not something that I usually forget to do (running codegen), while things like typos fail my CI very often.

Reply:
How often are you git pushing?

Reply:
I dunno. 5-20 times a day? 🤷

Reply:
Alright -- if you want to save those 25 - 100 seconds you can now:
```
export RERUN_LINT_SKIP=lint-codegen
```